### DISCUSSION TITLE
Mermaid subgraph extra spaces

### DISCUSSION THREAD

**Discussion by cumiller** [Doraku/DefaultDocumentation#65] [Q&A]
I am having trouble getting a mermaid diagram in the xml comments to render on Azure DevOps. DevOps does not seem to like that the subgraph end of line has two spaces after it.

ex.: 
::: mermaid
graph TD
A((Start))-->B
B-->A0

subgraph mySubgraph  
A0{?}
A0-- Yes-->A1
A1-->C
A0-- No -->B0
B0-->C
end

W-->X((end))
:::

After "mySubgraph" there are two spaces which cause DevOps' mermaid engine to error. The generated xml comments file does not have spaces after "mySubgraph". Is this something DefaultDocumentation is putting in?

**Comment by Doraku** [comment]
Yeah, this is the markdown syntax for a line break :/
```
this (no '  ' at the end)
text
```
render like this in markdown
> this (no '  ' at the end) text

Doing two line break instead of '  ' would generate a paragraphe, which is different from a normal line break.

Hum maybe I could add a custom balise like `<raw>` where the content is displayed as is without any reformatting. I will check if other documentation generators have something like this, if there is a precedent.

**Comment by cumiller** [comment]
Ok, I see. The mermaid commands are being seen as a text string and the double space is used to maintain the line breaks. DD would need a tag or delimiter to know that it is not a simple text string. Can the ::: ( and I've seen ```) used at the beginning and end of the mermaid diagram markup be used to signal to DD that it should be handled raw?

**Reply by Doraku** [reply]
hum actually the `<code>` element which produce the ``` wrapping should work fine if changed to write its content raw (currently it uses the same method as other text so it appends `  ` at the end of each line, I can change this). You can already declare a "language" attribute. If it works for you you would be able to do
```
<code language="mermaid">
graph TD
A((Start))-->B
B-->A0

subgraph mySubgraph
A0{?}
A0-- Yes-->A1
A1-->C
A0-- No -->B0
B0-->C
end

W-->X((end))
</code>
```
in your xml documentation which would produce in DefaultDocumentation (without the backslash, I just don't know how to escape the ` character in github)
```
\``` mermaid
graph TD
A((Start))-->B
B-->A0

subgraph mySubgraph
A0{?}
A0-- Yes-->A1
A1-->C
A0-- No -->B0
B0-->C
end

W-->X((end))
\```
```

But maybe having a `<raw>` element would still be the best.

**Reply by cumiller** [reply]
Hmmm, -me too. 

Are you saying that a <Code Language="Mermaid> section would render with the Mermaid delimeter (:::) and other <Code> sections (without Language=Mermaid) would be ignored and not get the "mermaid" handling?

That seems like it would work.

`<raw>` might be a more universal tag to bypass formatting for a section.

**Reply by cumiller** [reply]
What about simply using the ":::" to start and end a "raw" section? That way, there are no tags added that other systems, like Doxygen, won't understand.

Another way to go on this is to not use the line breaks in the xml to set the line breaks in the markdown. Most of the line breaks in the source code xml comments are simply there to fit the screen width in the development environment. It would be better in some ways if the block of text was one paragraph that was line wrapped for the final display rather than having embedded markdown newline characters.

[I realize I might be over-simplifying this - and, also, as a late-comer,  underestimate the challenges of xml to markdown]

**Reply by Doraku** [reply]
it's called xml documentation for a reason, it is xml based ^^" so it is easier to use xml tag for special handling as it is already parsed, no need to reparse the inner text. And as you said it would be a more universal solution:
```xml
<!-- ignore markdown reformating and copy as is the content, maybe still trimming the space at the start? -->
<raw>
::: mermaid
graph TD
A((Start))-->B
B-->A0

subgraph mySubgraph
A0{?}
A0-- Yes-->A1
A1-->C
A0-- No -->B0
B0-->C
end

W-->X((end))
:::
</raw>
```

> Are you saying that a <Code Language="Mermaid> section would render with the Mermaid delimeter (:::) and other \` sections (without Language=Mermaid) would be ignored and not get the "mermaid" handling?`

the default language is `csharp` but you can change it with the `language` attribute. You would still end up with \``` instead of ::: :/ so I guess it's a no. I never head of mermaid before but it seems pretty standard and widespread, but the ::: seems to be the Azure DevOps flavor? apparently gitlab uses ``` (which would work with the `code` element hijacking). For this reason I think I will add a \<raw> element, that way users can decide what they want.

**Reply by Doraku** [reply]
> Another way to go on this is to not use the line breaks in the xml to set the line breaks in the markdown. Most of the line breaks in the source code xml comments are simply there to fit the screen width in the development environment. It would be better in some ways if the block of text was one paragraph that was line wrapped for the final display rather than having embedded markdown newline characters.

I actually hesitated to do that before https://github.com/Doraku/DefaultDocumentation/issues/51#issuecomment-841100211

**Reply by Doraku** [reply]
> What about simply using the ":::" to start and end a "raw" section? That way, there are no tags added that other systems, like Doxygen, won't understand.

Sorry for multiple replies I am juste thinking of new stuff as I post ^^"
that's actually a good point (the part about tags not being understandable by other tools). What about custom attribute then? Do you think other tools wouldn't complain?
```xml
<code language="mermaid" delimiter=":::">
graph TD
A((Start))-->B
B-->A0

subgraph mySubgraph
A0{?}
A0-- Yes-->A1
A1-->C
A0-- No -->B0
B0-->C
end

W-->X((end))
</code>
```
or
```xml
<para formating="raw">
::: mermaid
graph TD
A((Start))-->B
B-->A0

subgraph mySubgraph
A0{?}
A0-- Yes-->A1
A1-->C
A0-- No -->B0
B0-->C
end

W-->X((end))
:::
</para>
```

**Reply by cumiller** [reply]
I am very carefully considering this issue. I am coming from the perspective of writing the docs info once and being able to output it in multiple ways (Markdown, HTML, PDF - using multiple tools) so I don't want to request something that breaks other tools.

The issue is when to apply newlines and when not to. In my xml comments, I have text sections (paragraph of text), list sections (text prefixed with "-") and mermaid diagrams. When I run it through Doxygen to create HTML output, I get a html file with the text sections shown as one paragraph (xml comment newlines removed) with html unnumbered lists for the "-" prefixed text sections. Blank lines between statements is maintained. The mermaid statements are all run together as one line-wrapped line (not supported, not rendered as a diagram). 

I tried `<code language="medium" delimiter=":::">` with Doxygen to see how it would render. In the HTML, it renders as a block of mono-spaced code.

`<para>` seems to have little affect on the output from Doxygen and from DD. In DD, since the ":::" delimiter is still present, the only issue is the double spaces. **I think `<para>` is the safest, least impact tag to use.**

**Reply by Doraku** [reply]
For a complete different approach, what about a global setting in DefaultDocumentation to handle line break in xml documentation as real line break (what is currently happening hence the `space space` postfix) or not (you would need to use `<para>` and `<br/>` to format your text, no more `space space` postfix for linebreak, this is what you actually need to do for intellisense).
Do you think that would work for you?
Actually this could be an attribute on each element too (`<summary>`, `<para>`, ...) so you can change the default setting locally.

**Reply by cumiller** [reply]
I do want to turn off line breaks for normal text regions. I want the documentation text region to flow to the width of the display and not follow the xml line breaks.

With the DD option enabled which turns off the postfix line breaks:

Would blank lines in the xml be ignored? Doxygen seem to figure out that a double CRLF is a desired break and puts a break in automatically.

What would need to be done for lists?

```
My List:
- First Item
- Second item which is really long and will eventually need to work wrap
so it does not go off of the screen
- Third Item

```
Each list item needs to start a line with “-“. Would a `<br/>` be required at the end of each item?

What would I need to do for my mermaid text region? It will not accept a `<br/>` on each line. I think I would still need a `<para formatting=”raw”>`  attribute for a mermaid section to keep the existing line breaks without putting in additional markup.

**Reply by Doraku** [reply]
> Would blank lines in the xml be ignored? Doxygen seem to figure out that a double CRLF is a desired break and puts a break in automatically.

It would write the documentation as it is in the xml documentation so the interpretation would be left to the markdown engine displaying your pages. Normally double CRLF is understood as a break line in markdown so it should works correctly.

> What would need to be done for lists?

Same as double CRLF, this syntax is directly understood by markdown engine so no need for `<br/>`. Basically it should only impact stuff like this:
```
two line of text with no two space postfix
and no list marker on the next line
```
should display as this in most (all?) markdown engine:
```
two line of text with no two space postfix and no list marker on the next line
```

**Reply by cumiller** [reply]
I guess what I'm thinking of is when Doxygen takes the mermaid text region and makes it one long string. I want to make sure the change to DD won't do that to lists or the mermaid diagrams.

**Reply by cumiller** [reply]
Do you have a release coming soon? I'd like to try your latest changes.

**Reply by Doraku** [reply]
Sorry the plan was to do a release last weekend but I was sick the full week and couldn't do much, and now I am away on holiday. I made a new beta release so you can easily test it through nuget (0.7.6-beta1).

**Reply by cumiller** [reply]
Thank you very much. Sorry to hear about you being ill. Hope you've had enough of a recovery to enjoy your holiday.

**Reply by cumiller** [reply]
The new release is looking good. Thanks for putting that out.

I have one issue. In my diagram text, I have some embedded <br/> tags to shorten long lines.

The xml file has `"<br/>"`. The markdown files output it as `"<br />"` (space added before the "/"). Is c# seeing that as an escape character and doing something special with it?   DevOps won't render it as a line break with the extra space.

In the XML:
```
::: mermaid
graph TD
A((Some really <br/>label that spans <br/>multiple lines))-->B
B-->C
:::
```

Gets written to the markdown file as:
```
::: mermaid
graph TD
A((Some really <br />label that spans <br />multiple lines))-->B
B-->C
:::
```

[Edited. I didn't notice the tags got lost when posted]

**Reply by cumiller** [reply]
Would you consider overriding the XElement.ToString() when writing out "br" elements?

MarkdownWrite WriteNodes()
`
 "br" when isPreview => _builder.Append(element.ToString()),
_ when isPreview => _builder,
_ => _builder.Append(element.Name.ToString() == "br" ? "<br/>" : element.ToString())
`

If element is "br" then write as custom string `"<br/>"` otherwise, write using ToString().

**Reply by Doraku** [reply]
Hum this is probably because of the framework XElement parsing and ToString overload. I will see if there is an easier way to preserve your tag spacing.

**Reply by cumiller** [reply]
Where are you at on this? Can we get <br/> written without inserting spaces?

**Reply by Doraku** [reply]
I just gave a quick look last week, sadly ICSharpCode.Decompiler return directly the documentation xml pre formatted losing whether you had a space or not and I don't think it is possible to change it, so I may have to replace this part completely. I am finally coming back home after 4 weeks of holidays/remote work so I will be in better disposition, sorry for the unproductive august month ^^"

**Reply by Doraku** [reply]
You should be able to declare
```
<code language="mermaid">
...
</code>
```
it will render as
````
```mermaid
...
```
````
which seem to work now on github at least.
